rfd/0128-aws-integration-tests.md

```diff
@@ -0,0 +1,191 @@
+that we don't introduce regressions by changing the permissions required by
+Teleport. To achieve this, GitHub action will be allowed to assume a set of
+roles that will simulate the minimum required permissions that each Teleport
+service requires to perform the required actions. Teleport services will be
+configured to assume these roles when running the integration tests - requires
+changes to the Teleport configuration - so we can test them in isolation.
+
+The AWS account will be configured with a set of roles that will be assumed by
+the GitHub actions pipeline. These roles will be configured with the minimum
+required permissions to run the integration tests. This will ensure that we
+don't introduce regressions by changing the permissions required by Teleport
+but those changes are not detected because another test requires the same
+permissions.
+
+AWS account configuration will be handled by IAC and will be stored in the
+Teleport Terraform repository. This will allow us to track changes to the configuration
+and ensure that we can revert them if needed while also allowing us to review
+the changes before they are applied. Each End-to-End test will run in a separate
+Job so we can configure the assumed role for each test. End-to-End action
+requires `id-token: write` permission to the GitHub OIDC provider to generate
+the OIDC token that will be used to authenticate with AWS API.
+
+## Tests
+
+This section describes the integration tests that will be added to the Teleport
+and what they will test.
+
+### Kubernetes Access
+
+Teleport supports the automatic discovery of Kubernetes clusters running on AWS.
+This is done by using the AWS API to discover the Kubernetes clusters running on
+the account that matches the configured labels and then using the Kubernetes API
+to forward the requests to the cluster.
+
+The first step of this process happens on the Teleport Discovery service and
+uses the AWS API credentials to poll the available clusters.
+If the discovery service finds a cluster that matches the configured labels, it
+will create a Kubernetes cluster - `kube_cluster` - object in the Teleport database.
+This object contains the information required to
+connect to the Kubernetes cluster. The second step of this process happens on
+the Kubernetes Service and uses the information from the `kube_cluster` object
+to generate a short-lived token that is used to authenticate with the Kubernetes
+API.
+
+The integration tests for Kubernetes access will test the following:
+
+- The discovery service can discover Kubernetes clusters running on AWS and
+  create the `kube_cluster` object in the Teleport database. If the object
+  already exists, it will be updated with the new labels.
+- The Kubernetes service receives the `kube_cluster` object and can generate
+  a short-lived token to authenticate with the Kubernetes API.
+- The Kubernetes service can forward the requests to the Kubernetes API.
```

Comment:
How will we be testing client connections? Can we use actual `kubectl` so simulate end users?

Same goes for the database access cc @smallinsky. I wonder if we can build a Docker image with all the client tools we need (kubectl, psql, mysql, etc.) and use it to connect, instead of relying on drivers. This way we will also be able to test specific client version if needed.

Reply:
`tsh` ships `tsh kubectl` so I was planning to use it to get secrets/services since both objects exist even if no cluster nodes exist.